src/lib/validationRules/index.ts

```diff
@@ -0,0 +1,162 @@
+  refreshTokenRule,
+  validGrantTypes,
+  validScope,
+  noUnknownFields,
+  defaultMaxAgeRule,
+  validApplicationType,
+  rightAuthenticationMethod,
+  validResponseType,
+  sameDomainForRedirectUris,
+  noLocalAndRemoteRedirectUris,
+  noUnsetClientUri,
+  // We omit localization validation as it is not supported by ESS or NSS (as of 2022-08-17) but leave it here for spec-compliance guidance.
+];
+
+export async function validateDocument(
+  jsonDocument: string,
+  rules: ValidationRule[]
+): Promise<ValidationResults> {
+  // try parsing first..
+  let clientIdDocument = {};
+  try {
+    clientIdDocument = JSON.parse(jsonDocument);
+  } catch {
+    return [
+      {
+        rule: {
+          type: "local",
+          name: "Document must be valid JSON",
+          description: "The document must be a valid JSON string.",
+        },
+        status: "error",
+        title: "Invalid JSON",
+        description: "The document could not be parsed to JSON.",
+        affectedFields: [],
+      },
+    ];
+  }
+
+  const validationPromises = rules.map(async (rule) => {
+    const results = await rule.check({ document: clientIdDocument });
+    return results.map((result) => ({
+      rule: rule.rule,
+      ...result,
+    }));
+  });
+
+  return (await Promise.all(validationPromises)).flat();
+}
+
+export async function requestRemoteValidation(
+  documentIri: string
+): Promise<ValidationResults> {
+  let response: Response;
+  try {
+    response = await await fetch(
+      `/api/validate-remote-document?documentIri=${documentIri}`,
+      {
+        method: "POST",
+      }
+    );
+  } catch (error) {
+    return [
+      {
+        status: "error",
+        affectedFields: [],
+        rule: { description: "", name: "", type: "remote" },
+        title: `Fetching resource was not successful`,
+        description: `An error ocured while trying to fetch the resource: ${error}`,
+      },
+    ];
+  }
+  let jsonResponse: any;
+  try {
+    jsonResponse = await response.json();
+  } catch (error) {
+    return [
+      {
+        status: "error",
+        affectedFields: [],
+        rule: { description: "", name: "", type: "remote" },
+        title: `Fetching resource was not successful`,
+        description: `The validation result could not be parsed.`,
+      },
+    ];
+  }
+
+  if (jsonResponse?.error) {
+    return [
+      {
+        status: "error",
+        affectedFields: [],
+        rule: { description: "", name: "", type: "remote" },
+        title: `Validation not successful`,
+        description: `The resource could not be validated: ${jsonResponse.error}`,
+      },
+    ];
+  }
+
+  if (!Array.isArray(jsonResponse.results)) {
+    return [
+      {
+        status: "error",
+        affectedFields: [],
+        rule: { description: "", name: "", type: "remote" },
+        title: `Fetching resource was not successful`,
+        description: `The validation result could not be parsed.`,
+      },
+    ];
+  }
```

Comment:
All of this is pretty unnecessary and would be covered by an integration test, that is, we should be able to assert that our own API hosted on our current server conforms to the API that we want, without testing explicit response values. (this is perhaps a bit on the side of "a little too paranoid" as a style of code)